### PR TITLE
 mds: reset heartbeat map at potential time-consuming places

### PR DESCRIPTION
**Pull request by ukernel**

### CODE REVIEW

File: src/mds/MDSContext.cc [hunk 1]
```diff
@@ -28,6 +28,9 @@ void MDSInternalContextBase::complete(int r) {
   assert(mds != NULL);
   assert(mds->mds_lock.is_locked_by_me());
   MDSContext::complete(r);
+
+  // avoid unhealthy heartbeat map when finish_contexts() processes large context list
+  mds->heartbeat_reset();
```
Comment (ukernel) [inline src/mds/MDSContext.cc]: Large context list are delayed messages (caused by frozen subtree). how large it is depends on how many messages were sent by clients while subtree was frozen. these messages need to be processed before dispatching new messages (to guarantee message order).  mds can't directly control how large the context list is (mds can control it indirectly by reducing max_export_size) . I can't figure out a better approach.

File: src/mds/Locker.cc [hunk 2]
```diff
@@ -2737,7 +2737,7 @@ void Locker::handle_client_caps(MClientCaps *m)
       mdcache->wait_replay_cap_reconnect(m->get_ino(), new C_MDS_RetryMessage(mds, m));
       return;
     }
-    dout(1) << "handle_client_caps on unknown ino " << m->get_ino() << ", dropping" << dendl;
+    dout(7) << "handle_client_caps on unknown ino " << m->get_ino() << ", dropping" << dendl;
```
Comment (batrick) [inline src/mds/Locker.cc]: Also needs to be a separate PR and the commit needs amended to note it fixes the issue. Sorry to be a PITA Zheng but this makes backporting easier and helps me get these fixes expedited.

File: src/mds/MDSContext.cc (hunk 1 above)
Comment (batrick) [reply src/mds/MDSContext.cc]: Thanks for explaining. What concerns me is that this hides the problem: the MDS is stuck processing a large context list and not doing anything else. This is precisely the type of scenario the heartbeat mechanism is designed to catch!

How about reducing the `max_export_size`? `1G` and even `100M` sounds too large. What are the drawbacks to breaking the exports into smaller chunks like `10M`?

File: src/mds/OpenFileTable.cc [hunk 3]
```diff
@@ -1045,10 +1045,14 @@ void OpenFileTable::_prefetch_dirfrags()
...
 
   MDSGatherBuilder gather(g_ceph_context);
+  int num_opening_dirfrags = 0;
   for (auto dir : fetch_queue) {
     if (dir->state_test(CDir::STATE_REJOINUNDEF))
       assert(dir->get_inode()->dirfragtree.is_leaf(dir->get_frag()));
     dir->fetch(gather.new_sub());
+
+    if (!(++num_opening_dirfrags % 100))
+      mds->heartbeat_reset();
```
Comment (batrick) [inline src/mds/OpenFileTable.cc]: Note: these don't bother me because we don't expect a non-active MDS to be responsive to clients.

Comment (ukernel) [reply src/mds/OpenFileTable.cc]: But if recovering mds does not send beacon to monitor, monitor may replace it with other mds

File: src/mds/MDSContext.cc (hunk 1 above)
Comment (ukernel) [reply src/mds/MDSContext.cc]: 10M is about 5000 inodes, that's too small

Comment (ukernel) [reply src/mds/MDSContext.cc]: 10M is about 5000 inodes, that's too small.  besides, we need to change the code that chooses export candidate. it's too inefficiency to freeze a large subtree, but only export small portion of it.

Comment (ukernel) [reply src/mds/MDSContext.cc]: large number of delayed messages is a generic problem. mds also may delay large number of messages during recovering. I don't think we can have better option for current code base (with the large mds_mutex)

**Comment by ukernel** [comment]
most change are moved to https://github.com/ceph/ceph/pull/23088

File: src/mds/Migrator.cc [hunk 4]
```diff
@@ -2593,6 +2582,14 @@ void Migrator::import_remove_pins(CDir *dir, set<CDir*>& bounds)
   }
 }
 
+class C_MDC_QueueContext : public MigratorContext {
+public:
+  std::list<MDSInternalContextBase*> contexts;
+  C_MDC_QueueContext(Migrator *m) : MigratorContext(m) {}
+  void finish(int r) override {
+    get_mds()->queue_waiters_front(contexts);
```
Comment (batrick) [inline src/mds/Migrator.cc]: Why do these need queued at the front?

File: src/mds/Migrator.cc [hunk 5]
```diff
@@ -169,17 +169,6 @@ void Migrator::dispatch(Message *m)
...
-class C_MDC_EmptyImport : public MigratorContext {
-  CDir *dir;
-public:
-  C_MDC_EmptyImport(Migrator *m, CDir *d) : MigratorContext(m), dir(d) {}
-  void finish(int r) override {
-    mig->export_empty_import(dir);
-  }
-};
-
-
```
Comment (batrick) [inline src/mds/Migrator.cc]: Dead code? Separate commit and PR please. (leaving this in here will mess up backports.)

File: src/mds/Migrator.cc (hunk 4 above)
Comment (batrick) [reply src/mds/Migrator.cc]: This will require that `MDSRank::finished_queue` be a `std::deque` which is not preferred to `std::vector` (in #23195 ). So, is it actually important to queue at the front?

File: src/mds/Migrator.cc [hunk 6]
```diff
@@ -1665,7 +1654,7 @@ uint64_t Migrator::encode_export_dir(bufferlist& exportbl,
 
 void Migrator::finish_export_dir(CDir *dir, mds_rank_t peer,
 				 map<inodeno_t,map<client_t,Capability::Import> >& peer_imported,
-				 list<MDSInternalContextBase*>& finished, int *num_dentries)
+				 list<MDSInternalContextBase*>& finished, unsigned& num_dentries)
```
Comment (batrick) [inline src/mds/Migrator.cc]: This isn't relevant to this PR. Needs to be a separate refactor PR that won't be backported.

File: src/mds/Migrator.cc (hunk 4 above)
Comment (ukernel) [reply src/mds/Migrator.cc]: yes, it's important.  to make sure contexts get processed in proper order

**Comment by batrick** [comment]
@ukernel please rebase

File: src/mds/MDSRank.cc [hunk 7]
```diff
@@ -818,14 +818,15 @@ void MDSRank::_advance_queues()
 {
   assert(mds_lock.is_locked_by_me());
 
-  while (!finished_queue.empty()) {
+  if (!finished_queue.empty()) {
     dout(7) << "mds has " << finished_queue.size() << " queued contexts" << dendl;
-    dout(10) << finished_queue << dendl;
-    decltype(finished_queue) ls;
-    ls.swap(finished_queue);
```
Comment (batrick) [inline src/mds/MDSRank.cc]: I think it would be preferable to keep `ls.swap` here to so that any buggy context which adds to the finished_queue doesn't cause an infinite loop.

File: src/mds/MDSRank.h [hunk 8]
```diff
@@ -307,6 +307,12 @@ class MDSRank {
       std::copy(v.begin(), v.end(), std::back_inserter(finished_queue));
       progress_thread.signal();
     }
+    void queue_waiters_front(MDSInternalContextBase::vec& ls) {
+      MDSInternalContextBase::vec v;
+      v.swap(ls);
+      std::copy(v.rbegin(), v.rend(), std::front_inserter(finished_queue));
```
Comment (batrick) [inline src/mds/MDSRank.h]: Wow. Okay this needs a comment explaining what situation this is addressing. You're basically reversing the context vector and putting that at the front of the deque? This is not what one glancing at `queue_waiters_front` would expect and it's easy to miss!

File: src/mds/Migrator.cc (hunk 4 above)
Comment (batrick) [reply src/mds/Migrator.cc]: Please document an example.

File: src/mds/MDSRank.h (hunk 8 above)
Comment (ukernel) [reply src/mds/MDSRank.h]: contexts in 'ls' do not get reversed.  If 'ls' is list, 'std::copy(v.rbegin(), v.rend(), std::front_inserter(finished_queue))' is equivalent to ’finished_queue.splice(finished_queue.begin(), ls)'

File: src/mds/MDSRank.cc (hunk 7 above)
Comment (ukernel) [reply src/mds/MDSRank.cc]: 'auto fin = finished_queue.front()' is required by queue_waiters_front()


if buggy context keeps adding another  buggy context to finished_queue,  it's infinite loop no matter keep "ls.swap" or not

**Comment by batrick** [comment]
And also needs a tracker ticket for backport.

**Comment by ukernel** [comment]
Fixes: http://tracker.ceph.com/issues/26858

File: src/mds/MDSRank.h (hunk 8 above)
Comment (batrick) [reply src/mds/MDSRank.h]: OKay got it

**Comment by batrick** [comment]
Just realized post-merging this that this PR may cause: http://tracker.ceph.com/issues/26869

Zheng, can you take a look?
